apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/V1/PooledSessionTests.cs

```diff
@@ -0,0 +1,294 @@
+using Google.Api.Gax.Testing;
+using Google.Protobuf;
+using Grpc.Core;
+using Moq;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Cloud.Spanner.V1.PoolRewrite.Tests
+{
+    // TODO: the mocking isn't ideal at the moment. We want to verify that only the calls we've set up are made, and only the right
+    // number of times, but Moq ends up being annoying in terms of the Settings property. This is definitely solvable, but needs
+    // more experience with Moq, I suspect.
+    public class PooledSessionTests
+    {
+        private static readonly SessionName s_sampleSessionName = new SessionName("project", "instance", "database", "session");
+        // In various tests we don't care whether things are evicted "on" or "after" a particular tick, so add one tick.
+        private static readonly TimeSpan OneTick = TimeSpan.FromTicks(1);
+
+        [Fact]
+        public void FromSessionName_BasicProperties()
+        {
+            var pool = new FakeSessionPool();
+
+            var pooledSession = PooledSession.FromSessionName(pool, s_sampleSessionName);
+
+            Assert.Same(s_sampleSessionName, pooledSession.SessionName);
+            Assert.False(pooledSession.ShouldBeEvicted);
+            Assert.False(pooledSession.RequiresRefresh);
+            Assert.False(pooledSession.ServerExpired);
+            Assert.Null(pooledSession.TransactionId);
+            Assert.Equal(TransactionOptions.ModeOneofCase.None, pooledSession.TransactionMode);
+        }
+
+        [Fact]
+        public void Expiry()
+        {
+            var pool = new FakeSessionPool();
+            var options = pool.Options;
+            var clock = (FakeClock) options.Clock;
+            var originalTime = clock.GetCurrentDateTimeUtc();
+
+            // If we evict sessions before we refresh them, things get weird.
+            Assert.True(options.IdleSessionRefreshDelay < options.PoolEvictionDelay);
+
+            var pooledSession = PooledSession.FromSessionName(pool, s_sampleSessionName);
+            Assert.False(pooledSession.ShouldBeEvicted);
+            Assert.False(pooledSession.RequiresRefresh);
+
+            // Detect the need for a refresh
+            clock.AdvanceTo(originalTime + options.IdleSessionRefreshDelay + OneTick);
+            Assert.False(pooledSession.ShouldBeEvicted);
+            Assert.True(pooledSession.RequiresRefresh);
+
+            // Detect the need for eviction
+            clock.AdvanceTo(originalTime + options.PoolEvictionDelay + OneTick);
+            Assert.True(pooledSession.ShouldBeEvicted);
+            Assert.True(pooledSession.RequiresRefresh);
+        }
+
+        // TODO: We'd quite like to check the next three methods for all operations (and maybe more?), but it's not easy to parameterize.
+        // Options:
+        // - Live with lots of duplication
+        // - Live without good unit test coverage
+        // - Have horribly complex parameterization
```

Comment:
Maybe?
// - Less horrible parameterization (Successful request method name, which would do everything in lines 97 to 100), a little bit of reflection to call that method and a Theory instead of a Fact receiving all the different successful request method names.

Reply:
We can do some things that way. Setting up the mock could be tricky though - I'll have to see.
I'll experiment while you're reviewing other PRs later :)

Reply:
If it helps kick start the experiments :)
```
        private async Task SuccessfulBeginTransaction(FakeSessionPool pool, PooledSession pooledSession)
        {
            var request = new BeginTransactionRequest();
            var response = new Transaction();
            pool.Mock.Setup(client => client.BeginTransactionAsync(request, It.IsAny<CallSettings>())).ReturnsAsync(response);
            await pooledSession.BeginTransactionAsync(request, 5, CancellationToken.None);
        }

        [Theory]
        [InlineData(nameof(SuccessfulBeginTransaction))]
        public async Task AutoRefreshOnSuccessfulRpc(string successfulRequest)
        {
            var pool = new FakeSessionPool();
            var options = pool.Options;
            var clock = (FakeClock)options.Clock;

            var originalTime = clock.GetCurrentDateTimeUtc();
            var pooledSession = PooledSession.FromSessionName(pool, s_sampleSessionName);

            var halfRefresh = TimeSpan.FromTicks(options.IdleSessionRefreshDelay.Ticks / 2);
            clock.Advance(halfRefresh);

            // Make a successful request
            // Some reflection here to call successfulRequest

            // The request will have extended the refresh time.
            Assert.Equal(clock.GetCurrentDateTimeUtc() + options.IdleSessionRefreshDelay, pooledSession.RefreshTimeForTest);
        }
```

Reply:
If only method groups could be passed as parameters to attributes we wouldn't even need the reflection :).